hera_cal/utils.py

```diff
@@ -1264,24 +1271,45 @@ def red_average(data, reds=None, bl_tol=1.0, inplace=False,
 
             # replace with new data
             if fed_container:
-                blkey = blg[0] + (pol,)
+                blkey = blk + (pol,)
                 data[blkey] = davg
                 flags[blkey] = favg
                 nsamples[blkey] = navg
 
             else:
-                blinds = data.antpair2ind(blg[0])
+                blinds = data.antpair2ind(blk)
                 polind = pols.index(pol)
                 data.data_array[blinds, 0, :, polind] = davg
                 data.flag_array[blinds, 0, :, polind] = favg
                 data.nsample_array[blinds, 0, :, polind] = navg
                 data.integration_time[blinds] = iavg
 
     # select out averaged bls
-    bls = [blg[0] + (pol,) for pol in pols for blg in reds]
+    bls = [blk + (pol,) for pol in pols for blk in red_bl_keys]
     if fed_container:
-        to_del = [bl for bl in data.keys() if bl not in bls]
-        del data[to_del], flags[to_del], nsamples[to_del]
+        # its much faster to assign a new dict then
+        # delete items from an existing dict for
+        # large data sets where number of red keys
+        # much smaller the original keys.
+        if not inplace:
+            new_data = {}
+            new_flags = {}
+            new_nsamples = {}
+            for bl in list(data.keys()):
+                if bl in bls:
+                    new_data[bl] = data[bl]
+                    new_flags[bl] = flags[bl]
+                    new_nsamples[bl] = nsamples[bl]
+            data = new_data
+            flags = new_flags
+            nsamples = new_nsamples
+        else:
+            # can't think of how to avoid this inplace.
+            for bl in list(data.keys()):
```

Comment:
You can delete multiple keys simultaneously, which is how I previously had this happen and which was a lot faster.

Reply:
Cool! Looks like I ended up pulling in some outdated code.